src/python/pants/base/specs.py

```diff
@@ -38,33 +90,104 @@ def __new__(cls, directory, name):
   def to_spec_string(self):
     return '{}:{}'.format(self.directory, self.name)
 
+  def matching_address_families(self, address_families_dict):
+    return self.address_families_for_dir(address_families_dict, self.directory)
+
+  class SingleAddressResolutionError(Exception):
+    def __init__(self, single_address_family, name):
+      super(SingleAddress.SingleAddressResolutionError, self).__init__()
+      self.single_address_family = single_address_family
+      self.name = name
+
+  def address_target_pairs_from_address_families(self, address_families):
+    """Return the pair for the single target matching the single AddressFamily, or error.
+
+    :raises: :class:`SingleAddress.SingleAddressResolutionError` if no targets could be found for a
+             :class:`SingleAddress` instance.
+    :return: list of (Address, Target) pairs with exactly one element.
+    """
+    single_af = assert_single_element(address_families)
+    addr_tgt_pairs = [
+      (addr, tgt) for addr, tgt in single_af.addressables.items()
+      if addr.target_name == self.name
+    ]
+    if len(addr_tgt_pairs) == 0:
+      raise self.SingleAddressResolutionError(single_af, self.name)
```

Comment:
Well, this is caught in [`_matching_addresses_for_spec()`](https://github.com/pantsbuild/pants/blob/6f51c0f56d8956f82df7f840d976531d64daf580/src/python/pants/engine/build_files.py#L247), and I was thinking this would make sense just containing the info that method needs to raise the prettier exception. Since this exception is noted in the docstring, I think raising it here is fine because callers will know to handle it.

Reply:
Ah to clarify, what I meant was something like this:
```python
raise self.SingleAddressResolutionError(single_af, 'Something went wrong with {}'.format(self.name))
```

I'm not sure what that message should be, but adding something more than `self.name`.

Reply:
I'm confused because in `_matching_addresses_for_spec()`, linked above, we specifically catch the `SingleAddressResolutionError`, and then re-raise with `_raise_did_you_mean(e.single_address_family, e.name)` (which produces the error message you see when you provide a target spec that doesn't exist). I think that provides all the context we need -- the `SingleAddressResolutionError` is just a way to get inputs for the `_raise_did_you_mean()` function. Am I missing something?

Reply:
Ah okay, that makes sense. I didn’t realize self.name has all this context.

Reply:
Well, it doesn't in itself, it's just a string, but the `_raise_did_you_mean()` method provides the context (and it now uses `raise_from` to preserve the full stacktrace as well).

Reply:
After discussion on slack, the internal exception class was made private (`_SingleAddressResolutionError`) in d6f66ab.